### PR TITLE
Fix #7258. Bug in SROA optimization

### LINKED ISSUE #7258: Reassign to an IR Variable Produces a Memory Leak
## Reporting a bug

- [X] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [ ] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

Starting with Numba 0.54, some of Bodo's tests are reporting memory leaks. This appears to be because we are failing to generate DEL nodes if an IR variable is assigned to twice in the same basic block. The IR generated doesn't appear to change, so it seems likely that Numba 0.54 is just the first version where we accurately detect these memory leaks. 
 
For context, we define a variable twice in the same basic block to achieve the behavior of `inplace=True` with a sort operation. We allocate a new array somewhere and then as a last step we assign the old IR variable to this new array.

I cannot produce a reproducer in just Numba, so I've included a segment of the IR before and after del nodes are added (omitting a bunch of unnecessary lines that add clutter). It should be possible to reproduce this with a pipeline extension, but if that's not necessary I want to avoid generating that reproducer.

Before adding del nodes:
```python
label 0:
    $2load_global.0.1090 = global(_get_rank: CPUDispatcher(<function get_rank at 0x10d7e2790>)) ['$2load_global.0.1090']
    rank_var.1093 = call $2load_global.0.1090(func=$2load_global.0.1090, args=[], kws=(), vararg=None, target=None) ['$2load_global.0.1090', 'rank_var.1093']
    $2load_global.0.1094 = global(_get_size: CPUDispatcher(<function get_size at 0x123627d30>)) ['$2load_global.0.1094']
    n_pes_var.1097 = call $2load_global.0.1094(func=$2load_global.0.1094, args=[], kws=(), vararg=None, target=None) ['$2load_global.0.1094', 'n_pes_var.1097']
    df = arg(0, name=df)                     ['df']
    col_ind.842 = const(int, 0)              ['col_ind.842']
    $2load_global.0.845 = global(bodo: <module 'bodo' from '/Users/nicholasriasanovsky/Documents/bodo/Bodo/bodo/__init__.py'>) ['$2load_global.0.845']
    $4load_attr.1.846 = getattr(value=$2load_global.0.845, attr=hiframes) ['$2load_global.0.845', '$4load_attr.1.846']
    $6load_attr.2.847 = getattr(value=$4load_attr.1.846, attr=pd_dataframe_ext) ['$4load_attr.1.846', '$6load_attr.2.847']
    $8load_method.3.848 = getattr(value=$6load_attr.2.847, attr=get_dataframe_data) ['$6load_attr.2.847', '$8load_method.3.848']
    $data0.894.1026 = call $8load_method.3.848(df, col_ind.842, func=$8load_method.3.848, args=[Var(df, test_sort_values.py:278), Var(col_ind.842, test_sort_values.py:278)], kws=(), vararg=None, target=None) ['$8load_method.3.848', '$data0.894.1026', 'col_ind.842', 'df']
    col_ind.851 = const(int, 1)              ['col_ind.851']

    ...(more IR omited)

    $data0.894.1026 = static_getitem(value=key_data.1168, index=0, index_var=None, fn=<built-in function getitem>) ['$data0.894.1026', 'key_data.1168']
    del key_data.1168                        []
    $data1.895.1027 = static_getitem(value=sort_data.1169, index=0, index_var=None, fn=<built-in function getitem>) ['$data1.895.1027', 'sort_data.1169']
    $arr.886.1024 = static_getitem(value=sort_data.1169, index=1, index_var=None, fn=<built-in function getitem>) ['$arr.886.1024', 'sort_data.1169']
...(more IR omited)
```

After adding del nodes:
```python
label 0:
    $data0.894.1026 = call $8load_method.3.848(df, col_ind.842, func=$8load_method.3.848, args=[Var(df, test_sort_values.py:278), Var(col_ind.842, test_sort_values.py:278)], kws=(), vararg=None, target=None) ['$8load_method.3.848', '$data0.894.1026', 'col_ind.842', 'df']
    del col_ind.842                          []
    del $8load_method.3.848                  []
    col_ind.851 = const(int, 1)              ['col_ind.851']

    ...(more IR omited but no del for $data0.894.1026 )

    $data0.894.1026 = static_getitem(value=key_data.1168, index=0, index_var=None, fn=<built-in function getitem>) ['$data0.894.1026', 'key_data.1168']
    del key_data.1168                        []
    $data1.895.1027 = static_getitem(value=sort_data.1169, index=0, index_var=None, fn=<built-in function getitem>) ['$data1.895.1027', 'sort_data.1169']
    $arr.886.1024 = static_getitem(value=sort_data.1169, index=1, index_var=None, fn=<built-in function getitem>) ['$arr.886.1024', 'sort_data.1169']
    del sort_data.1169                       []
...(more IR omited but no del for $data0.894.1026 )
    del $data1.895.1027                      []
    del $data0.894.1026                      []
    
```

The key takeaways are that we assigns `$data0.894.1026` twice, but the del is only generated once at the end.

### PR DESCRIPTION
**Pull request by sklam**
SROA opt inside lowering is incorrectly including variables that are assigned multiple times within the block.

## Reference an existing issue

fixes https://github.com/numba/numba/issues/7258
closes #7233

### CODE REVIEW

File: numba/tests/test_ssa.py [hunk 1]
```diff
@@ -493,3 +497,78 @@ def while_for(n, max_iter=1):
         self.assertPreciseEqual(while_for(10), while_for.py_func(10))
         # One phi?
         self.assertEqual(phi_counter, [1])
+
+
+class TestSROAIssues(MemoryLeakMixin, TestCase):
+    # This tests issues related to the SROA optimization done in lowering to
+    # reduce time spent in LLVM SROA pass. The optimization is related to SSA
```
Comment (stuartarchibald) [inline numba/tests/test_ssa.py:504]:
```suggestion
    # This tests issues related to the SROA optimization done in lowering, which
    # reduces time spent in the LLVM SROA pass. The optimization is related to SSA
```

File: numba/tests/test_ssa.py [hunk 2]
```diff
@@ -493,3 +497,78 @@ def while_for(n, max_iter=1):
...
+    # and tries to reduce the number of alloca statements for variables with
+    # only a single assignemnt.
+    def test_issue7258_multiple_assignment_post_SSA(self):
+        # This test adds a pass that will duplicate assignment statements to
+        # variables named "foobar".
+        # In reported issue, the bug will cause a memory leak.
```
Comment (stuartarchibald) [inline numba/tests/test_ssa.py:510]:
```suggestion
        # In the reported issue, the bug will cause a memory leak.
```

File: numba/tests/test_ssa.py [hunk 3]
```diff
@@ -493,3 +497,78 @@ def while_for(n, max_iter=1):
...
+        cloned = []
+
+        @register_pass(analysis_only=False, mutates_CFG=True)
+        class CloneFoobarAssignments(FunctionPass):
+            # A pass that clone variable assignments into "foobar"
```
Comment (stuartarchibald) [inline numba/tests/test_ssa.py:515]:
```suggestion
            # A pass that clones variable assignments into "foobar"
```

File: numba/tests/test_ssa.py [hunk 4]
```diff
@@ -493,3 +497,78 @@ def while_for(n, max_iter=1):
...
+                )
+                pm._finalized = False
+                # Insert the cloning pass after SSA
+                pm.add_pass_after(CloneFoobarAssignments, ReconstructSSA)
+                pm.finalize()
+                return [pm]
+
+        @njit(pipeline_class=CustomCompiler)
+        def udt(arr):
+            foobar = arr + 1  # this assigment will be cloned
```
Comment (stuartarchibald) [inline numba/tests/test_ssa.py:551]:
```suggestion
            foobar = arr + 1  # this assignment will be cloned
```

File: numba/tests/test_ssa.py [hunk 5]
```diff
@@ -493,3 +497,78 @@ def while_for(n, max_iter=1):
...
+        matched_lines = [ln for ln in lines
+                         if "foobar =" in ln and ln.lstrip().startswith('#')]
+        self.assertEqual(
+            len(matched_lines), 2,
+            "expected two assignment statements into 'foobar'",
+        )
+        self.assertEqual(
+            matched_lines[0], matched_lines[1],
+            "expected the two assignment statements to be the same",
+        )
```
Comment (stuartarchibald) [inline numba/tests/test_ssa.py:574]: Would the `numba.core.untyped_passes.PreserveIR` pass help with this? It saves the IR into the compile result metadata at point the pass is called.

File: numba/tests/test_ssa.py [hunk 6]
```diff
@@ -493,3 +497,78 @@ def while_for(n, max_iter=1):
...
+            foobar = arr + 1  # this assigment will be cloned
+            return foobar
+
+        arr = np.arange(10)
+        # Verify that the function works as expected
+        self.assertPreciseEqual(udt(arr), arr + 1)
+        # Verify that the expected statement is cloned
+        self.assertEqual(len(cloned), 1)
+        self.assertEqual(cloned[0].target.name, "foobar")
+        # Verify the Numba IR that the expected statement is cloned
```
Comment (stuartarchibald) [inline numba/tests/test_ssa.py:560]:
```suggestion
        # Verify in the Numba IR that the expected statement is cloned
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, this fixes the issue noted. I've made a few minor suggestions, once resolved, should be good to merge.

File: numba/core/lowering.py [hunk 7]
```diff
@@ -314,7 +314,14 @@ def _find_singly_assigned_variable(self):
                 if len(var_assign_map[var]) == 1:
                     # Usemap does not keep locally defined variables.
                     if len(var_use_map[var]) == 0:
-                        sav.add(var)
+                        # Ensure that the variable is not defined multiple times
+                        # the the block
+                        [defblk] = var_assign_map[var]
+                        assigns = [stmt for stmt in self.blocks[defblk].body
+                                   if isinstance(stmt, ir.Assign)
+                                   and stmt.target.name == var]
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/lowering.py:322]: ah yes

File: numba/tests/test_ssa.py (hunk 5 above)
Comment (sklam) [reply numba/tests/test_ssa.py:574]: oh right, i forgot about that

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes.

**Pull request merged**